### DISCUSSION TITLE
How to use PyThaiNLP without internet connection

### DISCUSSION THREAD

**Discussion by ranyaphat29** [PyThaiNLP/pythainlp#571] [Q&A] [answered]
Hi all,
I am using pythainlp for developing my module without internet connection after I installed the library, I found then some issue. What should I solve this problem ?

> Non-HTTP error occurred: HTTPSConnectionPool(host='raw.githubusercontent.com', port=443): Max retries exceeded with url: /PyThaiNLP/pythainlp-corpus/2.2/db.json (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7fa66fb3e9d0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution'))
> Cannot download corpus catalog from: https://raw.githubusercontent.com/PyThaiNLP/pythainlp-corpus/2.2/db.json
> Traceback (most recent call last):
>   File "/opt/conda/bin/uvicorn", line 8, in <module>
>     sys.exit(main())
>   File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 829, in __call__
>     return self.main(*args, **kwargs)
>   File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 782, in main
>     rv = self.invoke(ctx)
>   File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>   File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 610, in invoke
>     return callback(*args, **kwargs)
>   File "/opt/conda/lib/python3.7/site-packages/uvicorn/main.py", line 362, in main
>     run(**kwargs)
>   File "/opt/conda/lib/python3.7/site-packages/uvicorn/main.py", line 386, in run
>     server.run()
>   File "/opt/conda/lib/python3.7/site-packages/uvicorn/server.py", line 49, in run
>     loop.run_until_complete(self.serve(sockets=sockets))
>   File "/opt/conda/lib/python3.7/asyncio/base_events.py", line 587, in run_until_complete
>     return future.result()
>   File "/opt/conda/lib/python3.7/site-packages/uvicorn/server.py", line 56, in serve
>     config.load()
>   File "/opt/conda/lib/python3.7/site-packages/uvicorn/config.py", line 308, in load
>     self.loaded_app = import_from_string(self.app)
>   File "/opt/conda/lib/python3.7/site-packages/uvicorn/importer.py", line 20, in import_from_string
>     module = importlib.import_module(module_str)
>   File "/opt/conda/lib/python3.7/importlib/__init__.py", line 127, in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
>   File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 983, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
>   File "<frozen importlib._bootstrap_external>", line 728, in exec_module
>   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed

**Comment by wannaphong** [comment]
Can you check a ssl certificate verify? Is date present?

**Comment by bact** [comment, marked as answer]
Some data files that required by the lib is not immediately included with the lib.

When a required data file is about to be used for the first time, the lib will try to install it - by download it from the internet.
( https://raw.githubusercontent.com/PyThaiNLP/pythainlp-corpus/2.2/db.json is basically a list of data files and where they can be download from )

This is of course not a very good design, from the perspective of being self-contained and should be able to run without network.

Workaround: You can try to install those data files when you have internet connection. Once installed, they can be use without internet connection.

Please refer to https://github.com/pythainlp/pythainlp#command-line-interface for the command line interface that you can use to install data files.
